---
title: "Human SI transcriptomics"
author: "A. Gabor"
date: "8/19/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(biomaRt)
library(tidyverse)
```


The following script shows the problem with differential expression analysis and 
a potential fix to it. 


```{r import, message=FALSE, warning=FALSE, include=FALSE}
# SI transciptomics data: 
de_files <- list.files("./data/5-FU in vitro human/transcriptomics/SI/differential_expression/",full.names = T)
transcript_data <- tibble(files = de_files) %>% 
	mutate(fileID = gsub(".csv","",basename(files))) %>%
	separate(fileID,into = c("concentration","time"),sep = "_",remove = FALSE) %>%
	mutate(concentration = as.numeric(gsub("uM","",concentration) )) %>%
	mutate(time = as.numeric(gsub("h","",time) )) %>%
	mutate(trans_data = map(files,read_csv)) %>%
	unnest(trans_data) %>%
	select(-files) %>%
	rename(ensembl_gene_id= X1)
```


Number of genes are the same between conditions: 
```{r pressure, echo=FALSE}
transcript_data %>% group_by(fileID) %>%
	summarise(n_genes = length(unique(ensembl_gene_id))) %>% 
	print()
```

Checking the distribution of the log2FC data: 

```{r echo=FALSE}
transcript_data %>% 
	ggplot(aes(fileID,log2FoldChange)) + 
	geom_boxplot() +
	geom_point(data = filter(transcript_data,abs(log2FoldChange)>10),
							 aes(fileID,log2FoldChange),col="red")
```

There are some extreme values log2FC >50, which is probably a division by 0 problem, since 
2^50 ~ 1e15. We have emailed Daniela and Terezinha about this problem. 

Genes with log2Fc larger than 10 are: 

```{r echo=FALSE}
map_table <- AnnotationDbi::mapIds(org.Hs.eg.db::org.Hs.eg.db, ensembl, 'ENTREZID', 'ENSEMBL')

gene_symbols = AnnotationDbi::mapIds(org.Hs.eg.db::org.Hs.eg.db, keys = transcript_data$ensembl_gene_id, 'SYMBOL', 'ENSEMBL')
transcript_data$gene_symbols = gene_symbols

transcript_data %>% filter(abs(log2FoldChange) > 10 ) %>%
	select(fileID,concentration,time,ensembl_gene_id,gene_symbols,everything()) %>%
	arrange(ensembl_gene_id) %>%
	print() 
```


Principal components analysis:

```{r echo=FALSE}
pca_data <- transcript_data %>% 
	select(fileID,ensembl_gene_id,log2FoldChange) %>% 
	spread(ensembl_gene_id,log2FoldChange) %>% 
	column_to_rownames("fileID")
pca_res <- prcomp(pca_data)
tibble(X = pca_res$x[,1],Y = pca_res$x[,2],label = rownames(pca_data) ) %>%
	ggplot(aes(X,Y)) + geom_point() + geom_text(aes(label=label))
```

The PCA is also a bit strange: why the 48h samples are clustering away from the 
other 2 time-points? 



### Gene filtering
We traced back a gene with large fold change and it has 0 counts in all the 
three replica. This indicates that the preprocessing pipeline cannot handle genes 
with zero expression. 

We don't have access to the pipeline, so here we filter out genes that have zero 
counts in 2 out of the 3 replica in any condition (in the raw data).

```{r include=FALSE}
sample_meta <- read_tsv("./data/5-FU in vitro human/transcriptomics/SI/raw/metadata.txt")
raw_count <- read_tsv("./data/5-FU in vitro human/transcriptomics/SI/raw/gene_count.txt") %>%
	rename(gene_ensembl_id = X1)

# n_positive: in how many replica the gene was expressed
# n_replica: how many replica are there ( 3 for each)
raw_count_stats <- raw_count %>% gather(name,count,-gene_ensembl_id) %>%
	left_join(sample_meta, by = "name") %>% 
	group_by(gene_ensembl_id,condition) %>%
	mutate(n_positive = sum(count>0),
		   n_replica = n())
```

How many genes we loose if we remove those who are not expressed in a condition?
In the table: TRUE: remove; FALSE: keep. 

```{r}
# which genes to remove: 
# the ones that have n_positive <2 in at least 1 condition:
raw_count_stats <- raw_count_stats %>% 
	group_by(gene_ensembl_id) %>%
	summarise(to_remove = any(n_positive < 2)) 

raw_count_stats %>%
	pull(to_remove) %>% table() %>% print()

keep_genes <- raw_count_stats %>% filter(!to_remove) %>% pull(gene_ensembl_id)
```

we would remove 41.8k genes and keep 16552.

```{r}
filtered_transcript_data <- transcript_data %>% filter(ensembl_gene_id %in% keep_genes)
```


Now the boxplot looks much nicer: 

```{r echo=FALSE}
filtered_transcript_data %>% ggplot() + geom_boxplot(aes(fileID,log2FoldChange))
```

Checking pca:

```{r echo=FALSE}
pca_data_filtered <- filtered_transcript_data %>% 
	dplyr::select(fileID,ensembl_gene_id,log2FoldChange) %>% 
	spread(ensembl_gene_id,log2FoldChange) %>% 
	column_to_rownames("fileID")
pca_res_filtered <- prcomp(pca_data_filtered)
tibble(X = pca_res_filtered$x[,1],Y = pca_res_filtered$x[,2],label = rownames(pca_data_filtered) ) %>%
	ggplot(aes(X,Y)) + geom_point() + geom_text(aes(label=label))
```

The 48h data are still further away, but the points are spread more homogenously.

```{r,eval=FALSE}
write_rds(filtered_transcript_data, "./data/5-FU in vitro human/transcriptomics/SI/SI_transcriptomics_diffExp_filtered_data.rds")
```


