---
title: "Human SI and colon organids data analysis"
author: "A. Gabor"
date: "7/15/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(biomaRt)
library(tidyverse)
#BiocManager::install("dorothea")
library(dorothea)
```

## Intro

Here we compute the transcription factor and pathway activity of the 5-fu treated human in-vitro data.
This will give us an overview of the relevant pathways activated and responding to 
the drug, depending on time and concentration. 

## Issue with transcriptomics data
There is a problem in DE analysis: some genes with no expression gets very high
log2FC. Check the `human_SI_transcriptomics_issue` script to see details. 

We found that if we filter out the genes that have zero expression across more than 
one replica the issue is resolves. 



## Import and preprocess the transcriptomics data

First we import all the transcriptomics data of SI treated with 5-FU and Colon
treated with FU. 

```{r import, message=FALSE, warning=FALSE, include=FALSE}
# SI transciptomics data: 
SI_data <- read_rds("./data/5-FU in vitro human/transcriptomics/SI/SI_transcriptomics_diffExp_filtered_data.rds") %>%
	add_column(organ = "SI")
# Colon transciptomics data: 
Colon_data <- read_rds("./data/5-FU in vitro human/transcriptomics/colon/colon_transcriptomics_diffExp_filtered_data.rds") %>%
	add_column(organ = "colon")


transcriptomics_data <- bind_rows(SI_data,Colon_data) %>% mutate(sample_id = paste0(organ,"_",fileID))
```




## Calculate TF activity:

We use the Dorothea regulon, which contains the HGNC symbols of transcription factors and 
transcription factor targets. Therefore we need to map the Ensembl gene ids to HGNC. 
We use `BiomaRt` for this.

#### Gene name conversion
```{r include=FALSE}
# load dorothea regulons with the standard ABC confidence levels: 
regulons <- dorothea_hs %>%
  dplyr::filter(confidence %in% c("A", "B","C"))
# download mart
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
genes_dic <- getBM(  attributes=c("hgnc_symbol","ensembl_gene_id"), # "entrezgene_id"
  mart = mart)

#Finding missing genes: 
tr_genes <- transcriptomics_data$ensembl_gene_id %>% unique()
# number of genes found in Biomart
tr_genes_found <- sum(tr_genes %in% genes_dic$ensembl_gene_id)
# missing genes
tr_genes_notfound <- sum(!tr_genes %in% genes_dic$ensembl_gene_id)
```

- out of `length(tr_genes)` ensembl genes, we found `tr_genes_found`
- didn't find `tr_genes_notfound`

List of missing, but significantly changing genes: 
```{r}
# We are missing the following ENSEMBL ids: 
missing_genes = tr_genes[!tr_genes %in% genes_dic$ensembl_gene_id]
# around 10-15 genes are significantly different between conditions:
transcriptomics_data %>% 
	filter(ensembl_gene_id %in% missing_genes) %>% 
	filter(padj<0.05) %>%
	arrange(padj)
```


```{r include=FALSE}
n_missing_HGCN <- genes_dic %>% 
	filter(ensembl_gene_id %in% transcriptomics_data$ensembl_gene_id) %>%
	filter(nchar(hgnc_symbol)==0) %>% nrow()
```
`n_missing_HGCN` genes found by EnsemblID has no HGCN name. 

add the gene names to the transcription data
```{r}
transcriptomics_data <- transcriptomics_data %>%
	left_join(genes_dic, by = "ensembl_gene_id")
```

Some Ensembl id were not found, we removed those genes and also those that have no 
gene names. Averaged those transcripts that have the same gene name. 

```{r include=FALSE}
dorothea_data = transcriptomics_data %>%
	filter(!is.na(hgnc_symbol)) %>%
	filter(nchar(hgnc_symbol)!=0)

# in each condition, each gene (ensemble_gene_id) appears once:
dorothea_data %>% group_by(sample_id,ensembl_gene_id) %>% summarise(n_records = n()) %>%
	pull(n_records) %>% max()

# but more ensembl id corresponds to the same gene name
dorothea_data %>% group_by(sample_id,hgnc_symbol) %>% summarise(n_records = n()) %>%
	arrange(desc(n_records))

# we take the mean log2FC for these
dorothea_data_hgnc <- dorothea_data %>% 
	group_by(sample_id,hgnc_symbol) %>%
	summarise(log2FC = mean(log2FoldChange))
```


Check how many dorothea regulons are in the data
```{r}
regulons %>% 
	filter(target %in% dorothea_data$hgnc_symbol) %>% 
	nrow() %>%
	print()
nrow(regulons)
```

11360 of the 14033 TF target transcripts are found in the data. 

#### TF estimation

## sample-wise

run VIPER to estimate TF activity in each sample. 
We run the analysis on log2 FC, therefore we get TFs that are strongly up/down
regulated in treatment vs control. 
Due to the ranking of viper, TFs are ranked in each sample separaterly, therefore
samples cannot be compared here. 
```{r include=FALSE}
# remove those genes not found 
viper_input = dorothea_data_hgnc  %>%
	spread(sample_id,log2FC) %>%
	column_to_rownames("hgnc_symbol")

tf_activities_stat <- dorothea::run_viper(viper_input, regulons,
    options =  list(minsize = 5, eset.filter = FALSE, 
    cores = 1, verbose = TRUE, nes = TRUE),)
```

Here are the top 25 transcription factor for each condition. Red means upregulated, 
blue is downregulated. 

E2F1, E2F2 and E2F4 are consistently downregulated in each sample, interestingly, 
the strength of downregulation 
```{r ViperBarplot, echo=FALSE, fig.height=12, fig.width=10, message=FALSE, dpi=300}
tf_activities_stat_top25 <- tf_activities_stat %>%
    as.data.frame() %>% 
    rownames_to_column(var = "GeneID") %>%
	gather(condition,NES,-GeneID) %>%
    group_by(condition) %>%
    top_n(25, wt = abs(NES)) %>%
    arrange(NES)
    
ggplot(tf_activities_stat_top25,aes(x = reorder(GeneID, NES), y = NES)) + 
    geom_bar(aes(fill = NES), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 90, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Transcription Factors") +
	facet_wrap(~condition,scales = "free_y") + coord_flip()
ggsave("./figures/SI_TF_activity_top25.pdf",width = 10,height = 15)

```

We take the 5 top TF's from each conditions and check which are the genes that were used 
to determine their activities: 

```{r Volcano, warning=FALSE, dpi=300, fig.width=12, fig.height=10}
top_5_tf = tf_activities_stat %>%
    as.data.frame() %>% 
    rownames_to_column(var = "GeneID") %>%
	as_tibble() %>%
	gather(condition,NES,-GeneID) %>%
    group_by(condition) %>%
    top_n(5, wt = abs(NES)) %>%
    arrange(NES) %>% pull(GeneID) %>% unique()

regulons %>% filter(tf %in% top_5_tf) %>%
	inner_join(dorothea_data %>% rename(target = hgnc_symbol),by = "target") %>%
	group_by(sample_id,tf) %>%
	mutate(lfc_rank = n()-rank(abs(log2FoldChange))) %>% 
	mutate(label = ifelse(padj<0.05 & lfc_rank<5, target, "" ) ) %>%
	arrange(desc(abs(log2FoldChange))) %>%
	#filter(tf=="ZNF263") %>%
	ggplot(aes(log2FoldChange,-log10(pvalue))) +
	geom_point(aes(color=padj<0.05)) +
	ggrepel::geom_text_repel(aes(label=label)) +
	facet_grid(sample_id~tf,scales = "free")

```

 E2F4 transcirption factor was the most downregulated TF. Based on the prior knowledge, 
 this TF upregualtes most (218) of it's targets and downreagulates 14: 

```{r}
regulons %>% filter(tf =="E2F4") %>% select(mor) %>% table() %>% print()

```

The log2FC of it's targets 24hours after treatment: 
```{r,fig.height=12}
regulons %>% filter(tf %in% top_5_tf) %>%
	inner_join(dorothea_data %>% rename(target = hgnc_symbol),by = "target") %>%
	group_by(sample_id,tf) %>%
	mutate(lfc_rank = n()-rank(abs(log2FoldChange))) %>% 
	mutate(label = ifelse(padj<0.05 & lfc_rank<5, target, "" ) ) %>%
	arrange(desc(abs(log2FoldChange))) %>%
	filter(time==24)%>%
	filter(tf=="E2F4") %>%
	mutate(node_color = ifelse(padj<0.05,ifelse(mor==1,"sig_up","sig_down"),"non-sig")) %>%
	ggplot(aes(log2FoldChange,-log10(pvalue))) +
	geom_point(aes(color=node_color)) +
	ggrepel::geom_text_repel(aes(label=label)) +
	facet_grid(sample_id~tf) + scale_color_manual(values = c(sig_up = "blue",sig_down="red",`non-sig` = "grey"))
```
```{r,fig.height=12}
regulons %>% filter(tf %in% top_5_tf) %>%
	inner_join(dorothea_data %>% rename(target = hgnc_symbol),by = "target") %>%
	group_by(sample_id,tf) %>%
	mutate(lfc_rank = n()-rank(abs(log2FoldChange))) %>% 
	mutate(label = ifelse(padj<0.05 & lfc_rank<5, target, "" ) ) %>%
	arrange(desc(abs(log2FoldChange))) %>%
	filter(time==24)%>%
	filter(tf=="TP53") %>%
	mutate(node_color = ifelse(padj<0.05,ifelse(mor==1,"sig_up","sig_down"),"non-sig")) %>%
	ggplot(aes(log2FoldChange,-log10(pvalue))) +
	geom_point(aes(color=node_color)) +
	ggrepel::geom_text_repel(aes(label=label)) +
	facet_grid(sample_id~tf) + scale_color_manual(values = c(sig_up = "blue",sig_down="red",`non-sig` = "grey"))
```

### comparable results: 


```{r}
# skip_one_merge
#
# returns a dataframe, where each column contains all, but the i-th column of
# the data from the original dataframe  
# input: data.frame
#
skip_one_merge <- function(df){
	
	dat = tibble()
	
	rb <- lapply(seq(ncol(df)),function(i){
		tmp = tibble(col = unlist(df[,-i]))
		colnames(tmp) = colnames(df)[i]
		return(tmp)
	} )
	lrb <- bind_cols(rb) %>% as.data.frame()
	
	rownames(lrb) = paste(rep(rownames(df),ncol(df)-1), rep(1:(ncol(df)-1),each=nrow(df)),sep = "_")
	return(lrb)
}

# data frame with the original data
viper_input = dorothea_data_hgnc  %>%
	spread(sample_id,log2FC) %>%
	column_to_rownames("hgnc_symbol")

# reference: each column contains all the expression data, therefore, when TFs are 
# ranked by the log2FC of their targets, all samples are considered: 
reference_log2fc = skip_one_merge(viper_input)

viper_w_ref = bind_rows(viper_input,reference_log2fc)
tf_activities_stat_w_ref <- dorothea::run_viper(viper_w_ref, regulons,
    options =  list(minsize = 5, eset.filter = FALSE, 
    cores = 1, verbose = TRUE, nes = TRUE),)
```

```{r ViperBarplot, echo=FALSE, fig.height=12, fig.width=10, message=FALSE, dpi=300}
global_tf_activities_stat_top25 <- tf_activities_stat_w_ref %>%
    as.data.frame() %>% 
    rownames_to_column(var = "GeneID") %>%
	gather(condition,NES,-GeneID) %>%
    group_by(condition) %>%
    top_n(25, wt = abs(NES)) %>%
    arrange(NES)
    
ggplot(global_tf_activities_stat_top25,aes(x = reorder(GeneID, NES), y = NES)) + 
    geom_bar(aes(fill = NES), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 90, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Transcription Factors") +
	facet_wrap(~condition,scales = "free_y") + coord_flip()

ggsave("./figures/SI_TF_activity_top25_global.pdf",width = 10,height = 15)
```

We take the 5 top TF's from each conditions and check which are the genes that were used 
to determine their activities: 

```{r Volcano, warning=FALSE, dpi=300, fig.width=5, fig.height=8}
top_5_tf = tf_activities_stat_w_ref %>%
    as.data.frame() %>% 
    rownames_to_column(var = "GeneID") %>%
	as_tibble() %>%
	gather(condition,NES,-GeneID) %>%
    group_by(condition) %>%
    top_n(5, wt = abs(NES)) %>%
    arrange(NES) %>% pull(GeneID) %>% unique()

regulons %>% filter(tf %in% top_5_tf) %>%
	inner_join(dorothea_data %>% rename(target = hgnc_symbol),by = "target") %>%
	group_by(sample_id,tf) %>%
	mutate(lfc_rank = n()-rank(abs(log2FoldChange))) %>% 
	mutate(label = ifelse(padj<0.05 & lfc_rank<5, target, "" ) ) %>%
	arrange(desc(abs(log2FoldChange))) %>%
	#filter(tf=="ZNF263") %>%
	ggplot(aes(log2FoldChange,-log10(pvalue))) +
	geom_point(aes(color=padj<0.05)) +
	ggrepel::geom_text_repel(aes(label=label)) +
	facet_grid(sample_id~tf,scales = "free")

```
```{r}
 tf_global <- tf_activities_stat_w_ref %>%
    as.data.frame() %>% 
    rownames_to_column(var = "GeneID") %>%
	gather(condition,NES_global,-GeneID) 


tf_local <- tf_activities_stat %>%
    as.data.frame() %>% 
    rownames_to_column(var = "GeneID") %>%
	gather(condition,NES_local,-GeneID) 


full_join(tf_local,tf_global,by = c("GeneID", "condition")) %>%
	ggplot() +geom_point(aes(NES_local,NES_global,col=condition))

```

```{r}



znf_targets <- regulons %>% filter(tf =="ZNF263") %>% pull(target)
viper_input %>% rownames_to_column("geneID") %>%
	select(`SI_10uM_72h`,geneID) %>%
	
	mutate(rank = rank(`SI_10uM_72h`)) %>%
	sample_n(313)%>%
	#filter(geneID %in% znf_targets) %>%
	pull(rank) %>% hist(.,n=100)

```
